### PR TITLE
KEYCLOAK-19230 Calling getTopLevelGroups is slow inside GroupLDAPStorageMapper#getLDAPGroupMappingsConverted

### LINKED ISSUE #14820: Calling getTopLevelGroups is slow inside GroupLDAPStorageMapper#getLDAPGroupMappingsConverted
### Description

See https://issues.redhat.com/browse/KEYCLOAK-19230 and the related discussion and comments on that issue.

### Discussion

_No response_

### Motivation

_No response_

### Details

_No response_

### PR DESCRIPTION
**Pull request by bohmber**
Calling getTopLevelGroups is slow inside GroupLDAPStorageMapper#getLDAPGroupMappingsConverted

Closes #14820 

[KEYCLOAK-19230](https://issues.redhat.com/browse/KEYCLOAK-19230)

### CODE REVIEW

File: federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/membership/group/GroupLDAPStorageMapper.java
Comment (mposolda) [inline federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/membership/group/GroupLDAPStorageMapper.java:390]: Minor: It will be good if calls, which are called often and affect the performance use the regex pattern instead of directly calling `String.split` . I am pretty sure you remember your recent PR for the LdapDn performance improvement, which was related exactly to this :-)

File: federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/membership/group/GroupLDAPStorageMapper.java
Comment (mposolda) [inline federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/membership/group/GroupLDAPStorageMapper.java:380]: Instead of introducing this method, will it be better if the `KeycloakModelUtils.findGroupByPath` is changed this way by default? If it is better in your environment, it will be most likely for other environments as well.

IMO it does not makes sense that just GroupLDAPMapper uses this "optimized" version when all other Keycloak places use the "default" version, which is most likely worse for performance.

**Review by mposolda — DISMISSED** [review]
@bohmber Thanks for the PR! Few points:

- The PR introduces new method to the model API and hence I am adding @hmlnarik as a reviewer to doublecheck if this matches with the requirements for the new store and doublecheck if implementation in the new stores (Map store) is ok
- Added 2 comments regarding possible optimization of the SQL in the JPA layer to call only single SQL in "getGroupByName" instead of two separate queries (one just for return ID and another for lookup group by this ID).
- Will be good to doublecheck if we have DB index by parent and name. This can add even more performance points. Could you doublecheck if such index exists in the DB?
- I wonder that we can change `KeycloakModelUtils.findGroupByPath` to be more optimal and leverage the new API method for getGroupByName and parent.
- I've added few more comments below regarding cache invalidations and some other minor things...

When you do some more optimizations, can you please share in the JIRA some results of the performance before this change and after this change? If possible, it will be ideal to share:
- Count of LDAP top level groups
- Do you use some custom "group path" for the Group LDAP mapper or just top level groups?
- How the login of the user performs before and after this change? How many groups your typical user has?

Thanks!

@hmlnarik WDYT about the points above and about this PR in general?

File: model/jpa/src/main/java/org/keycloak/models/jpa/entities/GroupEntity.java
Comment (bohmber) [reply model/jpa/src/main/java/org/keycloak/models/jpa/entities/GroupEntity.java:40]: I'm fine with rename `getGroupByName` to `getGroupByNameAndParent`.  I guess the lookup of the id is needed because the entity must fetched from the cache if there is an entry in the cache instead of db. I'm doing the lookup like the realmById and realmByName

File: model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RealmCacheSession.java
Comment (bohmber) [reply model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RealmCacheSession.java:270]: That is a good point. I will review the GroupEvents. Thanks

File: federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/membership/group/GroupLDAPStorageMapper.java
Comment (bohmber) [reply federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/membership/group/GroupLDAPStorageMapper.java:390]: If the pattern is short there is a fast path in `String.split` without a Pattern. I need to check if a static Pattern will improve performance in this case.

File: federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/membership/group/GroupLDAPStorageMapper.java
Comment (bohmber) [reply federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/membership/group/GroupLDAPStorageMapper.java:380]: That was my wish as well. But there is a handling of groups with / in the name in `KeycloakModelUtils.findGroupByPath` Any good idea how to solve this?

**Comment by bohmber** [comment]
> * Will be good to doublecheck if we have DB index by parent and name. This can add even more performance points. Could you doublecheck if such index exists in the DB?

At least there is a `@UniqueConstraint(columnNames = {"REALM_ID", "PARENT_GROUP", "NAME"})`

File: model/jpa/src/main/java/org/keycloak/models/jpa/entities/GroupEntity.java
Comment (mposolda) [reply model/jpa/src/main/java/org/keycloak/models/jpa/entities/GroupEntity.java:40]: Yes, true. Good point with the lookup by id. Thanks for checking the other places in the codebase, which is something, which I should do as well before making this comment :-)

Comment (mposolda) [reply model/jpa/src/main/java/org/keycloak/models/jpa/entities/GroupEntity.java:40]: +1 for rename to `getGroupByNameAndParent`

File: model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RealmCacheSession.java
Comment (bohmber) [reply model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RealmCacheSession.java]: Changed the cache invalidation code

**Comment by mposolda** [comment]
@bohmber Still interested in this PR? If yes, please comment here.

There will be at least rebase needed and align with latest Keycloak codebase, which changed since this PR was sent. But possibly some more things will be needed to be updated in this PR. But before going further, I would like to know if you are still interested in this PR or not anymore.

**Comment by bohmber** [comment]
> @bohmber Still interested in this PR? If yes, please comment here.
> 

Thanks @mposolda 
Yes I'm interested in the PR. I know I need to rebase and update the pr. Currently we have almost the same change for rh-sso 7.3 and 7.6 running in production and we are happy with the results. I need at least 2-3 week until I can update the pr busy with a 2nd 7.6 migration

**Comment by bohmber** [comment]
@mposolda pr updated. I'm looking forward to your feedback

**Comment by ah-ping-luk** [comment]
I also want to vote for this fix and in my company, with very large AD, there are many users (especially those senior staffs) are assigned to 100+ LDAPS groups. We also suffer from this issue.

**Comment by bohmber** [comment]
Only LDAP is affected. There should be an internal RH issue with more details. It's hard to reproduce you need Group modifications in ldap, a high group count and many calls to getTopLevelGroups. If the cache is invalidated it will hammer the database with Group calls. https://issues.redhat.com/browse/KEYCLOAK-19230 has some information and the internal support case should have some data from a jfr session.

**Comment by ah-ping-luk** [comment]
Actually, it is not difficult to reproduce it…

You just need to create 100 ldap groups. Create an ldap user. Assign the user into the groups created. 

Config user federation with the ldap group mapper.

Login with the ldap user created, you will find it takes more than 10 seconds to login (time taken depends on cpu speed) for the first time (or cases that the user not loaded into cache).

**Review by mhajas — CHANGES_REQUESTED** [review]
The approach for fixing the performance issue seems correct, however, I added a few comments. 

What I am missing in this PR are some tests for the new method you added, for example, in model testsuite. 

@bohmber I am aware it is a long time since you sent a PR, but we would like to wrap this up in the following weeks. Would you be able to look at the comments and tests?

**Comment by bohmber** [comment]
> @bohmber I am aware it is a long time since you sent a PR, but we would like to wrap this up in the following weeks. Would you be able to look at the comments and tests?
@mhajas give me a few days then I can spend some time on this pr again

File: server-spi/src/main/java/org/keycloak/models/RealmModel.java
Comment (bohmber) [reply server-spi/src/main/java/org/keycloak/models/RealmModel.java]: Without this how it works with cacheSession?

File: server-spi/src/main/java/org/keycloak/storage/group/GroupLookupProvider.java
Comment (bohmber) [reply server-spi/src/main/java/org/keycloak/storage/group/GroupLookupProvider.java]: Order changed. 
Default implementation added.
Having the parentGroup is solving a lot of problems related to groupPath. I'm sceptical if I can see all possible cases in a more generic way.

File: model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/entities/GroupNameQuery.java
Comment (bohmber) [reply model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/entities/GroupNameQuery.java]: done

File: federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/membership/group/GroupLDAPStorageMapper.java
Comment (bohmber) [reply federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/membership/group/GroupLDAPStorageMapper.java:379]: This code is for the groupPath in the GroupLDAPStorageMapper.
`.property().name(GroupMapperConfig.LDAP_GROUPS_PATH)
                .label("Groups Path")
                .helpText("Keycloak group path the LDAP groups are added to. For example if value '/Applications/App1' is used, "
                		+ "then LDAP groups will be available in Keycloak under group 'App1', which is child of top level group 'Applications'. "
                		+ "The default value is '/' so LDAP groups will be mapped to the Keycloak groups at the top level. "
                		+ "The configured group path must already exists in the Keycloak when creating this mapper.")`

File: server-spi/src/main/java/org/keycloak/models/RealmModel.java
Comment (mhajas) [reply server-spi/src/main/java/org/keycloak/models/RealmModel.java]: `session.groups()` is `cacheSession` in the background. See:
https://github.com/keycloak/keycloak/blob/28a18bf7b94bde30c16abd04ac248fa347f1857d/model/legacy-private/src/main/java/org/keycloak/storage/datastore/LegacyDatastoreProvider.java#L143-L151

All methods in `RealmModel` that do not work with the realm entity are just redirecting the call to `session.area().correctMethod()`. We would like to eventually get rid of these methods so it is cleaner.

Comment (bohmber) [reply server-spi/src/main/java/org/keycloak/models/RealmModel.java]: thanks now I understand

File: federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/membership/group/GroupLDAPStorageMapper.java
Comment (bohmber) [reply federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/membership/group/GroupLDAPStorageMapper.java:381]: @mhajas ok we can start with your approach. It might be good enough to extend the validation in GroupLDAPStorageMapper that a group with a '/' in the Name is not good for performance

Comment (mhajas) [reply federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/membership/group/GroupLDAPStorageMapper.java:381]: I like documenting this behaviour in the tooltip/docs better than refusing configs with a slash in the name. As I said, there may be a use case where someone wants to use a group with a slash in its name. If the validation refuses such config, they cannot do it. The only thing we can do with this is to create a new discussion/issue where we discuss how to get rid of groups with `/` in the name completely from the whole Keycloak. We can't change it only in one part of the code. @sguilhen Any thoughts on this?

File: server-spi/src/main/java/org/keycloak/models/RealmModel.java
Comment (bohmber) [reply server-spi/src/main/java/org/keycloak/models/RealmModel.java]: @mhajas removed `RealmModel#getGroupByName`

File: federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/membership/group/GroupLDAPStorageMapper.java
Comment (bohmber) [reply federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/membership/group/GroupLDAPStorageMapper.java:382]: @mhajas I will try to support '/' in groupNames. Give me a few days. I'm a little bit busy with other tasks

Comment (sguilhen) [reply federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/membership/group/GroupLDAPStorageMapper.java:382]: @mhajas I agree with your view that we should not refuse groups with slash in the name as that is currently allowed. As you pointed out, we can discuss whether this makes sense or not in a separate issue, but for the scope of this PR this is a behavior I would not change.

Comment (bohmber) [reply federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/membership/group/GroupLDAPStorageMapper.java]: @mhajas support for '/' in groupNames added. Can you give me some hints where and which kind of tests I should create. I'm not familar with the testing infrastructure of keycloak. There is no real test of KeycloakModelUtils.findGroupByPath beside testsuite/federation/ldap

File: server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java
Comment (mhajas) [inline server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java:767]: Is there some reason to keep the old `findGroupByPath` implementation? Can we replace that method everywhere?

File: federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/membership/group/GroupLDAPStorageMapper.java
Comment (mhajas) [reply federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/membership/group/GroupLDAPStorageMapper.java]: Thank you for the changes @bohmber! 

Regarding tests, we have some tests where we use `findGroupByPath`. For example, https://github.com/keycloak/keycloak/blob/main/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/federation/ldap/LDAPGroupMapperSyncTest.java

The changes you added can be tested without LDAP. Not sure if we need any additional LDAP tests as ^LDAPGroupMapperSyncTest should cover the changes you made to the group mapper. In my opinion, what we need is the following:
1. Tests for the newly added method to GroupLookupProvider (`GroupModel getGroupByName(RealmModel realm, GroupModel parent, String name)`). We can create some groups, ideally with conflicting names under more parents and make sure the method returns the correct one. Another test can test the invalidation: call `getGroupByName` which should cache the result, then try to remove the group and make sure the follow-up call will return `null`. This can be added here: https://github.com/keycloak/keycloak/blob/main/testsuite/model/src/test/java/org/keycloak/testsuite/model/group/GroupModelTest.java see https://github.com/keycloak/keycloak/blob/main/testsuite/model/README.md on how to run
2. Then we should also add some tests for the new `findGroupByPath` method. It can be added to the same test class as 1. Let me know if you need any help, we can work together on that.

**Review by mhajas — COMMENTED** [review]
Thank you for the changes @bohmber. I added two comments. I like the new implementation of `findGroupByPath` therefore I would like to investigate the option to remove the old implementation and replace it with the new method. It would require quite a lot of changes (the old method is used in many places) and there are different arguments to the new method. Let me know if you want any help with the replacement.

**Comment by sguilhen** [comment]
> Thank you for the changes @bohmber. I added two comments. I like the new implementation of `findGroupByPath` therefore I would like to investigate the option to remove the old implementation and replace it with the new method. It would require quite a lot of changes (the old method is used in many places) and there are different arguments to the new method. Let me know if you want any help with the replacement.

@mhajas wouldn't it preferable to handle this re-factoring as a follow up issue/PR?

**Comment by mhajas** [comment]
> > Thank you for the changes @bohmber. I added two comments. I like the new implementation of `findGroupByPath` therefore I would like to investigate the option to remove the old implementation and replace it with the new method. It would require quite a lot of changes (the old method is used in many places) and there are different arguments to the new method. Let me know if you want any help with the replacement.
> 
> @mhajas wouldn't it preferable to handle this re-factoring as a follow up issue/PR?

@sguilhen Thank you for the suggestion. Both make sense. I am not sure whether we should call it re-factoring I see it more like optimization though. But anyway, if @bohmber prefers to not make so many changes here we can do it in a follow-up. In that case, I would mark the old method with `@Deprecated` annotation with a comment to use the new method.

File: server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java
Comment (bohmber) [reply server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java:767]: Backwards compatibility?

Comment (bohmber) [reply server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java:767]: Good question I'm not sure to replace the method everywhere in the first step.

**Comment by bohmber** [comment]
@mhajas Can you comment my latest changes?

File: server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java
Comment (mhajas) [reply server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java:767]: I don't think we need to keep backward compatibility for the util method in the `server-spi-private` module. Let's keep this as is currently and do a follow-up where we remove the old method later. I created an issue tracker for it: https://github.com/keycloak/keycloak/issues/23344

Comment (sguilhen) [reply server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java:767]: +1 to handle this in separate issue

**Review by mhajas — DISMISSED** [review]
Thank you for adding the tests @bohmber! I added one more commit with an additional test case and also updated one of your test cases to do the search in a separate transaction. I am approving now. Let's wait for @sguilhen for his feedback.

**Review by sguilhen — APPROVED** [review]
It does look good to me now, @bohmber ! Thanks for the effort you've put into it.

**Comment by mhajas** [comment]
We are waiting for @mposolda whether the @keycloak/core team wants to have a look.

**Comment by bohmber** [comment]
@mhajas Thanks for the Test. Found a misunderstanding of `searchForGroupByNameStream` in the default Method `GroupLookupProvider#getGroupByName` and a missing branch in `MapGroupProvider#getGroupByName`

**Review by mhajas — CHANGES_REQUESTED** [review]
Thanks for the additions @bohmber, however, I am not following the latest changes, could you elaborate more on why you did it?

File: server-spi/src/main/java/org/keycloak/storage/group/GroupLookupProvider.java
Comment (bohmber) [reply server-spi/src/main/java/org/keycloak/storage/group/GroupLookupProvider.java:47]: @mhajas that was my first expectation as well. But I removed the getGroupByName in the MapGroupProvider to test the default impl and the test was failing. If you read the javadoc 
`Returns the group hierarchy with the given string in name for the given realm.
 For a matching group node the parent group is fetched by id (with all children) and added to the result stream.
 This is done until the group node does not have a parent (root group)`

in this case you are getting for example a stream with

<name> (group1 (toplevel group))
<parent1><name> (group2 (child of parent))
<parent3><parent2><name> (group3 (child of parent2..))

and we are only interested in the toplevel groups that means we need to filter out the groups with fetched parent group.

Comment (mhajas) [reply server-spi/src/main/java/org/keycloak/storage/group/GroupLookupProvider.java:47]: I see. Thanks for the explanation, I missed that. Do you think it would make sense to have the default implementation to use `realm.getTopLevelGroups`? 

```java
default GroupModel getGroupByName(RealmModel realm, GroupModel parent, String name) {
    return (parent == null ? realm.getTopLevelGroupsStream() : parent.getSubGroupsStream())
            .filter(groupModel -> groupModel.getName().equals(name)).findFirst().orElse(null);
}
```

Comment (bohmber) [reply server-spi/src/main/java/org/keycloak/storage/group/GroupLookupProvider.java:47]: @mhajs that's a much better idea. If you have a lot of toplevel Groups it might be slower. But that is less confusing and that is the default impl. If someone reports it's slow it can implemented in the class more efficient.

Comment (mhajas) [reply server-spi/src/main/java/org/keycloak/storage/group/GroupLookupProvider.java]: Exactly, thanks @bohmber

**Review by mhajas — APPROVED** [review]
Thank you for the contribution @bohmber! I am approving and merging now

**Comment by ahus1** [comment]
Awesome, thank you for everyone involved in preparing and review this!

**Comment by bohmber** [comment]
@mhajas @sguilhen @mposolda @ahus1 Thanks to everyone. From the title the pr looks small. But in a case the toplevel group cache becomes invalid every call for group mapping will hammer the database for all groups for each group in the ldap response. Depending on the request count and group count you will get really bad response times quicky. Now there is a direct mapping between name and group and you will not affected when a group has added/changed.

**Comment by ahus1** [comment]
To those watching this issue: The related PR has been backported to 22.0.5

**Pull request merged**
